packages/contracts/contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol

```diff
@@ -969,6 +1083,33 @@ contract OVM_ExecutionManager is iOVM_ExecutionManager, Lib_AddressResolver {
             bytes memory
         )
     {
+        uint256 messageValue = _nextMessageContext.ovmCALLVALUE;
+        // If there is value in this message, we need to transfer the ETH over before switching contexts.
+        if (messageValue > 0) {
+            // Handle out-of-intrinsic gas consistent with EVM behavior -- the subcall "appears to revert"
+            if (gasleft() < CALL_WITH_VALUE_INTRINSIC_GAS) {
+                return (false, hex"");
+            }
+
+            // If not out of intrinsic gas, we guarantee that that intrinsic gas is reserved for consumption by the EM and OVM_ETH.
+            _gasLimit = Math.min(
```

Comment:
I think this logic needs a few more detailed comments. Even after having gone through it a bunch of times last week I still have to remind myself why this is correct.

Reply:
Agreed. Specifically this `min` is not super intuitive. This is my understanding:
- This function is called with `_gasLeft` which is the amount of gas the user desires to make the subcall with.
- We must reserve at least `INTRINSIC_GAS` to make sure that we transfer back the ETH if the subcall reverts.
- When you specify a `gasLimit` for a call that is GREATER than `gasLeft()` the behavior of the EVM is to just pass the `gasLeft()` down to the subcall & ignore the fact that you were trying to call with way too much gas. (is that right?)
- The `Math.min(...)` has two cases:
  - a) `gasLeft() - intrinsic gas` is a smaller value, in which case we set the `gasLimit` to that lower value.
    - This can **never** overflow because of the check above `gasleft() < CALL_WITH_VALUE_INTRINSIC_GAS`
  - b) `_gasLimit` is a smaller value, in which case we have more than enough time gas to return the ETH
- So in either case we EITHER: 1) have enough gas to return the ETH, or 2) the entire call will revert.

This is definitely confusing and worth a solid comment.

Reply:
I pushed more comments, can you tell me what you think?

Reply:
It looks good to me! If I were to super nit this then I would suggest that you add the comment:
```
                gasleft() - CALL_WITH_VALUE_INTRINSIC_GAS // cannot overflow because of the gasLimit check above
```

That said, personally I'm OK with just shipping this.

Reply:
added